### PR TITLE
argocd-autopilot cli added & tar binary name fix

### PR DESCRIPTION
**Pull request by dirien**
Signed-off-by: Engin Diri engin.diri@mail.schwarz
## Description

- argocd-autopilot cli added
- Argocd-autopilot cli is compressed with the filename containing the arch.

image

The current implementation of the arkade extraction is assuming, that the file in the tar is named as the CLI command. This is not always the case. And we should not rely on this. So instead, when we encounter a tar, we use the filename in the tar for the further copy command to the .arkade/bin folder.

## Motivation and Context

https://github.com/alexellis/arkade/issues/406

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
```bash
$ docker container run --rm -v $(pwd):/workspace -ti golang:1.16.3-alpine sh
# cd /workspace
# go build
# ./arkade get argocd-autopilot
```
image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

- [ ] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

**Comment by developer-guy** [comment]
LGTM

**Comment by developer-guy** [comment]
PTAL @alexellis @Waterdrips 🙏

**Comment by alexellis** [comment]
Please can you give an example of the before/after behaviour?

I.e. list two URLs (one that worked) and (one which did not) and what happened

Then what happens after.

Thanks

**Comment by dirien** [comment]
HI Alex,

Try this URL -> https://github.com/argoproj-labs/argocd-autopilot/releases/download/v0.1.6/argocd-autopilot-linux-amd64.gz

The binary in the tar is argocd-autopilot-linux-amd64. Arkade is not able to copy i as it expects the toolname (without arch or smt.) the toolname in the temp folder. But its untared with the arch name.

image

**Comment by alexellis** [comment]
Hi @dirien we don't need all these commits, so please could you squash them to result in a single commit?

https://www.youtube.com/watch?v=8j0H6urZ-bU

Alex

File: pkg/get/download.go [hunk 1]
```diff
@@ -40,22 +40,11 @@ func Download(tool *Tool, arch, operatingSystem, version string, downloadMode in
...
-		if strings.Contains(strings.ToLower(operatingSystem), "mingw") && tool.NoExtension == false {
-			outFilePath += ".exe"
-		}
-
-		if strings.HasSuffix(downloadURL, "tar.gz") || strings.HasSuffix(downloadURL, "tgz") {
-			untarErr := archive.Untar(archiveFile, outFilePathDir)
+		var tarFileName string
+		if strings.HasSuffix(downloadURL, "gz") || strings.HasSuffix(downloadURL, "tar.gz") || strings.HasSuffix(downloadURL, "tgz") {
+			var untarErr error
+			tarFileName, untarErr = archive.Untar(archiveFile, outFilePathDir)
```
Comment (alexellis) [inline pkg/get/download.go:47]: Unfortunately, there can often be more than one file within an archive. Helm is an example of that, and at times there is also a README and license file.

Can you explain your approach here, or your algorithm for determining the binary name?

Comment (dirien) [reply pkg/get/download.go:47]: Do you have an arkade example for me to test the helm case?

Comment (alexellis) [reply pkg/get/download.go:47]: I suggest that you look at the tests to see which filenames mention an archive like .zip or .tar.gz

You also didn't answer my question yet:

> Can you explain your approach here, or your algorithm for determining the binary name?

Alex

**Comment by dirien** [comment]
I will fix this differently. With the help of BinaryTemplate and URLTemplate. Will open a new PR.

**Comment by alexellis** [comment]
Please could you propose an issue so that we can discuss the design and approach, before you work on the PR? Overall, I hope this approach will save you time.

Thanks,

Alex

**Pull request closed without merging**